### PR TITLE
New cursor-based implementation of grapheme clusters

### PR DESCRIPTION
**Pull request by raphlinus**
This is a draft of the implementation for #21. It should be useful both for random access (determining the previous or next boundary from any starting offset) and for dealing with noncontiguous string representations such as ropes.

I'd probably want to add doc examples. Also, testing of the cursor-based API is inadequate, it's only tested through the existing iterators (which are implemented in terms of cursors). But I think it's complete enough to take a look at.

In its current form, this patch is _not_ suitable for purely streaming use cases (as described in #7), as incomplete chunk input may cause a `PreContext` return even when enough input has already been provided. It would be possible to add more state to the cursor, at a cost of some additional complexity. However, this implementation should at least avoid quadratic behavior when moving the cursor through a sequence of flags.

### CODE REVIEW

**Comment by raphlinus** [comment]
Note: this fixes #19 (which, I believe, had an incorrect forward iterator in the old codebase, because it was left in the Emoji state after the first fitzpatrick skin tone modifier).

File: src/grapheme.rs [hunk 1]
```diff
@@ -85,351 +83,469 @@ impl<'a> Graphemes<'a> {
...
+/// Cursor-based segmenter for grapheme clusters.
+#[derive(Clone)]
+pub struct GraphemeCursor {
+    offset: usize,  // current cursor position
+    len: usize,  // total length of the string
+    is_extended: bool,
+    state: GraphemeState,
+    cat_before: Option<GraphemeCat>,  // category of codepoint immediately preceding cursor
+    cat_after: Option<GraphemeCat>,  // category of codepoint immediately after cursor
+    pre_context_offset: Option<usize>,
```
Comment (Manishearth) [inline src/grapheme.rs]: nit: needs more comments on how this works

File: src/grapheme.rs [hunk 2]
```diff
@@ -85,351 +83,469 @@ impl<'a> Graphemes<'a> {
...
+        (GC_Regional_Indicator, GC_Regional_Indicator) => Regional,  // GB12, GB13
+        (_, _) => Break,  // GB999
+    }
+}
 
-        self.cat = if take_curr {
-            idx = idx + self.string[idx..].chars().next().unwrap().len_utf8();
-            None
+impl GraphemeCursor {
+    /// Create a new cursor. The string and initial offset are given at creation
```
Comment (Manishearth) [inline src/grapheme.rs]: nit: Examples on how the whole cursor API is supposed to be used?

File: src/grapheme.rs (hunk 1 above)
Comment (raphlinus) [inline src/grapheme.rs]: Done

File: src/grapheme.rs [hunk 3]
```diff
@@ -85,351 +83,469 @@ impl<'a> Graphemes<'a> {
...
-            }
-        }
+#[derive(PartialEq, Eq)]
+enum PairResult {
+    NotBreak,  // definitely not a break
+    Break,  // definitely a break
+    Extended,  // a break if not in extended mode
+    CheckCrlf,  // a break unless it's a CR LF pair
+    Regional,  // a break if preceded by an even number of RIS
+    Emoji,  // a break if preceded by emoji base and extend
```
Comment (raphlinus) [inline src/grapheme.rs]: I changed this to read "a break iff not in extended mode" which I think is more precise. What you suggest is the converse of what I originally said, but both are true.

**Comment by Manishearth** [comment]
Merged! Thanks so much for working on this!

**Pull request merged**